[BEAM-8347]: Consistently advance UnboundedRabbitMqReader watermark

drobert opened the pull request:
Originally reported in [BEAM-8347](https://issues.apache.org/jira/browse/BEAM-8347?focusedCommentId=16953004&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16953004) and discussed on [StackOverflow](https://stackoverflow.com/a/58421999/796064) the UnboundedRabbitMqReader has surprising watermarking behavior.

Notably:
- Watermarks only advance if new messages come through. It's easily the case that no new messages come through for some long periods of time, preventing downstream windows from closing.
- The implementation of `getWatermark` runs counter to suggested documentation in the `UnboundedSource` documentation.

From the `getWatermark` javadoc:

> watermark] can be approximate. If records are read that violate this guarantee, they will be considered late, which will affect how they will be processed. ...
> 
> However, this value should be as late as possible. Downstream windows may not be able to close until this watermark passes their end.
> 
> For example, a source may know that the records it reads will be in timestamp order. In this case, the watermark can be the timestamp of the last record read. For a source that does not have natural timestamps, timestamps can be set to the time of reading, in which case the watermark is the current clock time.

From the above, a Rabbit/queue-based client should be able to safely expect records read are in timestamp order. This implies the watermark should be the timestamp of the last record read (or 'now' if the records have no timestamp).

The current implementation uses the following logic for setting the watermark on a checkpoint:
```java
        currentTimestamp = new Instant(delivery.getProperties().getTimestamp());
        if (currentTimestamp.isBefore(checkpointMark.oldestTimestamp)) {
          checkpointMark.oldestTimestamp = currentTimestamp;
        }
```
That is, the timestamp remains stuck at the *oldest* timestamp. The watermark effectively never advances until a call to `finalizeCheckpoint` is triggered, which often does not happen with low volumes of messages.

This PR makes the following significant changes to the implementation:
1. When no new messages are available, the watermark advances to wall clock time
2. When a new message is processed, watermark advances to delivery time (if present), else current time

Monotonically advancing the watermark should be safe, even with 'split'/multi-consumers running against a single queue. Because it's a queue, processing for any given shard should be in consistent FIFO order. The client implementation also preserves in-order semantics of fetched messages. 

This implementation is very similar to the implementation used in [AmqpIO](https://github.com/apache/beam/blob/master/sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java#L308) and analogous to the KafkaIO [ProcessingTimePolicy](https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/TimestampPolicyFactory.java#L112)

NOTE: I am having a hard time finding prior art on how to test an UnboundedSource.

R: @jbonofre 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -376,16 +389,11 @@ public void finalizeCheckpoint() throws IOException {
       this.current = null;
       this.checkpointMark = checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark();
-      try {
-        connectionHandler = new ConnectionHandler(source.spec.uri());
```
Fixing a separate bug here. Originally, the constructor created the `connectionHandler` instance, which was closed inside `close()`. However `start()` created one as well, but masked the member variable with a local instance, so the channel/connection actually used for processing was not the one managed by the lifecycle of the class.

This is a semantic difference though, as now validation of the connection handler happens during `start()` rather than during construction. If there's a strong preference, I can restore this and instead modify `start()` to use the existing connection handler

kennknowles replied:
My impression is that your change here is an improvement.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -479,10 +488,10 @@ public boolean advance() throws IOException {
-        if (currentTimestamp.isBefore(checkpointMark.oldestTimestamp)) {
-          checkpointMark.oldestTimestamp = currentTimestamp;
-        }
+        Date deliveryTimestamp = delivery.getProperties().getTimestamp();
```
The meat of the change. Note I'm having trouble finding any documentation on exactly how `getProperties().getTimestamp()` is populated. I'm falling back to `Instant.now()` as a safeguard (which should actually be safe).

drobert commented:
I'd love to get some eyes and feedback on this before the 2.17.0 release is cut.
R: @jbonofre (you're listed in OWNERS)
R: @iemejia (you approved my last PR)

Thanks.

kennknowles commented:
Even though 2.17.0 has been cut, there is still time to cherrypick. Sorry for the large review delay.

kennknowles commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -376,16 +389,11 @@ public void finalizeCheckpoint() throws IOException {
     @Override
     public Instant getWatermark() {
-      return checkpointMark.oldestTimestamp;
+      return checkpointMark.latestTimestamp;
```
Ideally, the watermark would be informed by the RabbitMQ broker's backlog. The goal is for the watermark to be a good prediction of the fact "no future records will have timestamps before the watermark's timestamp".

My impression is that you have made the watermark advance more reliably, in good ways, but will perhaps result in a large amount of data being marked late. What do you think?

drobert replied:
I'm not sure I follow. 

>  The goal is for the watermark to be a good prediction of the fact "no future records will have timestamps before the watermark's timestamp".

In this implementation, the timestamp is effectively processing time. The timestamp is the delivery timestamp (which is how this was implemented even before this PR), which I believe is when the consumer picked up the message, not when it was enqueued. The queue is FIFO (we should probably add explicit documentation that priority queues are unsupported).

As noted in the PR description, the AmqpIO and certain Kafka timestamp watermarking implementations directly use 'now', so there's prior art here.

> when there's not a lot of messages (but in that case, we can imagine to update the watermark before the finalize checkpoint

This use case is what prompted this PR originally. 

> when we use an existing exchange (instead of creating one)

I don't see how this is relevant. Messages are pulled from a single queue, regardless of how the exchange was declared. 

> As it depends of the use case

drobert commented:
> Instead of changing the overall watermark advance behavior, it would make sense to provide for user to define the way to update the watermark.

I have a few concerns here, though this idea did cross my mind.

1. I'm having a hard time envisioning another reasonable watermarking strategy. The only real context is 'delivery time', which as noted will be FIFO, so I'm not sure what situation would cause the strategy in this PR to be inappropriate
2. Pluggable strategies feel like they ought to be a separate PR. It's a larger problem, and I'd set out to solve the more immediate, which is that the current strategy fails with low message counts.

jbonofre commented:
@drobert yeah, I agree. And actually, I also had hard times to think about an efficient watermark strategy in these IOs (jms, amqp, rabbitmq, mqtt, ...).  I used the same approach in these IOs: the watermark is the oldest timestamp before finalizing the checkpoint. For instance, in JmsIO checkpoint, when the checkpoint is finalized I do:

```
        message.acknowledge();
        Instant currentMessageTimestamp = new Instant(message.getJMSTimestamp());
        snapshot.updateOldestPendingTimestampIf(currentMessageTimestamp, Instant::isAfter);
```

So, the watermark advances with the oldest pending message timestamp.

The finalize checkpoint happens when the bundle is completed, so, I think it's consistent.

Right ?

drobert commented:
@jkff I looked at the PubSub implementation as well and opted against it *for now* for a few reasons:
1. It has an open ticket for the watermark not advancing with low volumes (https://issues.apache.org/jira/browse/BEAM-7322) which is the issue I set out to solve here for rabbit
2. It's fairly complicated relative to what I have here for rabbit. Given the difficulties in testing watermarking for an unbounded source, I didn't relish making this a blocker.
3. Rabbit has different semantics than pubsub. I actually think the processing time watermark is correct and simple for rabbit.

I'm very open to exploring better watermarking approaches in the future, but I think the current one is broken and I'd like to get a fix in first and improvements to it in second, if that makes sense.

jkff commented:
@drobert I totally agree that it makes sense to do a stop-the-bleeding fix first. Watermark logic in unbounded sources that don't have good native watermarks has been an issue for a very long time; if one were to attack this problem in a generic way, it would be best to start with a design and discuss it on dev@ - certainly outside the scope of this PR.

drobert commented:
Quick note: I may be seeing a problem with this implementation locally. This isn't mergeable due to a need to rebase anyway, but there may be an issue updating the watermark when no messages come in (DirectRunner reports "Processing stuck in step (read from rabbit) for at least 05m00s without outputting or completing in state process"). Looking.

drobert commented:
Update: after running some tests, the migration to the pull-based API in https://github.com/apache/beam/pull/9900 proves to be unsafe for this use case. I misread or misunderstood the semantics but it seems `channel.basicGet` blocks indefinitely, as opposed to returning immediately if there are no messages. This means if no new messages come in for a long period of time, the Runner will detect that the reader is 'stuck' since it's been on a single blocking call for so many minutes. Basically, we need to revert to a push-based-but-blocking API for this.

*However* that is a separate issue from the watermarks advancing. I don't think that broken merge should block this PR once rebased. I'll have a separate PR to fix the other up shortly.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -530,6 +543,10 @@ public boolean advance() throws IOException {
         // we consume message without autoAck (we want to do the ack ourselves)
         GetResponse delivery = channel.basicGet(queueName, false);
         if (delivery == null) {
+          current = null;
```
Notable change since the last round of reviews: The documentation for `Source`'s `getCurrent`, `getCurrentRecordId` and `getCurrentTimestamp` include the following:

>      * @throws NoSuchElementException if the reader is at the beginning of the input and {@link
>     *     #start} or {@link #advance} wasn't called, or if the last {@link #start} or {@link
>     *     #advance} returned {@code false}.

This states that if `advance()` returns `false`, subsequent calls to these should throw `NoSuchElementException`. As such, I'm explicitly setting these to `null`.

drobert commented:
@jkff This has been rebased and should be ready to merge.

I did make one notable change since the last review which is I now explicit set `current`, `currentTimestamp` and `currentRecordId` to `null` per this javadoc in `Source` for each corresponding getter:

```java
@throws NoSuchElementException 
   ... if the last {@link #start} or {@link#advance} returned @code false}.
```

jbonofre commented:
@jkff sure, I do a final pass and I merge ;) Thanks !
